crates/ty_ide/src/hover.rs

```diff
@@ -20,7 +20,21 @@ pub fn hover(db: &dyn Db, file: File, offset: TextSize) -> Option<RangedValue<Ho
     }
 
     let model = SemanticModel::new(db, file);
-    let ty = goto_target.inferred_type(&model).map(HoverContent::Type);
+    let ty =
+        goto_target.inferred_type(&model).map(|ty| match ty {
+            Type::KnownInstance(KnownInstanceType::TypeVar(typevar)) => typevar
+                .bind_pep695(db)
+                .map_or(HoverContent::Type(ty, None), |typevar| {
+                    HoverContent::Type(
+                        Type::NonInferableTypeVar(typevar),
+                        Some(typevar.variance(db)),
+                    )
+                }),
+            Type::TypeVar(typevar) | Type::NonInferableTypeVar(typevar) => {
+                HoverContent::Type(ty, Some(typevar.variance(db)))
+            }
+            _ => HoverContent::Type(ty, None),
+        });
```

Comment:
Nit, I think I'd move the mapping to `HoverContent` to line 52. 

```suggestion
    let ty = goto_target.inferred_type(&model);
    
    ...
    
    let mut contents = Vec::new();
    contents.extend(match ty {
        Type::KnownInstance(KnownInstanceType::TypeVar(typevar)) => typevar
            .bind_pep695(db)
            .map_or(HoverContent::Type(ty, None), |typevar| {
                HoverContent::Type(
                    Type::NonInferableTypeVar(typevar),
                    Some(typevar.variance(db)),
                )
            }),
        Type::TypeVar(typevar) | Type::NonInferableTypeVar(typevar) => {
            HoverContent::Type(ty, Some(typevar.variance(db)))
        }
        _ => HoverContent::Type(ty, None),
    });
```